File: app/src/main/java/edu/temple/myapplication/MainActivity.java
package edu.temple.myapplication;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import android.os.Bundle;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    String name;
    private final String NAME_KEY = "name";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        if (getSupportActionBar() != null)
            getSupportActionBar().setTitle("My Dashboard");

        if (savedInstanceState != null)
            name = savedInstanceState.getString(NAME_KEY);
        else
            name = "NO NAME ENTERED";

        findViewById(R.id.saveButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                name = ((EditText) findViewById(R.id.editText)).getText().toString();
            }
        });

        findViewById(R.id.showButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(MainActivity.this, name.toUpperCase(), Toast.LENGTH_SHORT).show();
            }
        });
    }

    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putString(NAME_KEY, name);
    }
}

File: app/src/main/java/edu/temple/activitystate/MainActivity.kt
package edu.temple.activitystate

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.*
import androidx.lifecycle.ViewModelProvider

class MainActivity : AppCompatActivity() {

    // A companion object is used here to store references
    // Think of this as something similar to a static reference
    // You can use 'const val', but constants have global namespace
    companion object {
        val KEY_PERSON = "personKey"
    }

    var person: Person? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val mainViewModel = ViewModelProvider(this).get(MainViewModel::class.java)

        // Fetch info from saveInstanceState, but only if it's not-null
        // argument will be null first time activity loads, and will continue
        // to be null if we do not store items in bundle argument of
        // onSaveInstanceState()
        savedInstanceState?.getString(KEY_PERSON)?.run {
            person = Person(this)
        }

        val nameEditText = findViewById<EditText>(R.id.editTextTextPersonName)
        val imageView = findViewById<ImageView>(R.id.imageView)

        findViewById<ToggleButton>(R.id.toggleButton).setOnCheckedChangeListener { _, p1 ->
            mainViewModel.setImageId(
                if (p1) R.drawable.cat2 else R.drawable.cat1
            )
        }


        // Observe changes to LiveData object stored in ViewModel
        // and update UI when changes occur
        mainViewModel.getImageId().observe(this) {
            imageView.setImageResource(it)
        }

        findViewById<Button>(R.id.saveButton).setOnClickListener {
            person = Person(nameEditText.text.toString())
        }

        findViewById<Button>(R.id.showButton).setOnClickListener {
            person?.run {
                Toast.makeText(this@MainActivity, "Name: $name", Toast.LENGTH_SHORT).show()
            }
        }
    }

    // Alternative to onCreate()
    // You can use one, but probably not both - Nothing will break, it's just inefficient
    override fun onRestoreInstanceState(savedInstanceState: Bundle) {
        super.onRestoreInstanceState(savedInstanceState)
        savedInstanceState.getString(KEY_PERSON)?.run {
            person = Person(this)
        }
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        // If person is not-null, then execute the scope function
        person?.run{
            outState.putString(KEY_PERSON, name)
        }
    }
}